src/net_processing.cpp

```diff
@@ -386,8 +386,18 @@ struct CNodeState {
     const CBlockIndex* pindexBestKnownBlock{nullptr};
     //! The hash of the last unknown block this peer has announced.
     uint256 hashLastUnknownBlock{};
-    //! The last full block we both have.
-    const CBlockIndex* pindexLastCommonBlock{nullptr};
+
+    //! The last full block we both have (per chainstate).
+    //!
+    //! This is namespaced by chainstate to allow syncing two separate chainstates
+    //! simultaneously from a single peer.
+    //!
+    //! Nota bene: this is contingent on the ChainstateManager not destructing
+    //! any CChainState objects which were in use at any point (e.g. a background
+    //! validation chainstate which has completed) until the end of
+    //! init.cpp:Shutdown(), else we'll have bad pointers here.
+    std::map<const CChainState*, const CBlockIndex*> chainstate_to_last_common_block = {};
```

Comment:
I think we can do this without adding `Chainstate` pointers to `CNodeState` which i would really prefer.

What I would suggest is that we nuke `pindexLastCommonBlock` entirely and compute it on demand instead (the performance overhead should be minimal). Then we don't need to keep track of it per chainstate but rather only compute it on demand for a given chain. I have a branch that should illustrate what i mean: https://github.com/dergoegge/bitcoin/commits/2022-10-assumeutxo-netproc (last commit puts the loop over the chains inside of `FindNextBlocksToDownload`, not sure if that works out)

Reply:
[rehashing a discussion from CoreDev]

This is a good idea, but should be considered separately as it will require benchmarking and more thorough analysis to ensure that it e.g. doesn't open some kind of DoS vector. I'm not familiar with why these values were cached in the first place, and so "un-caching" them is a design choice I'd rather not conflate here with assumeutxo.